monai/utils/nvtx.py

```diff
@@ -0,0 +1,105 @@
+# Copyright 2020 - 2021 MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Decorators and context managers for NVIDIA Tools Extension to profile MONAI components
+"""
+
+from collections import defaultdict
+from functools import wraps
+from typing import Any, Optional
+
+from monai.utils import optional_import
+
+_nvtx, _ = optional_import("torch._C._nvtx", descriptor="NVTX is not installed. Are you sure you have a CUDA build?")
+
+__all__ = ["Range"]
+
+
+class Range:
+    """
+    A decorator and context manager for NVIDIA Tools Extension (NVTX) Range for profiling.
+    When used as a decorator it encloses a specific method of the object with an NVTX Range.
+    When used as a context manager, it encloses the runtime context (created by with statement) with an NVTX Range.
+
+    Args:
+        name: the name to be associated to the range
+        method: (only when used as decorator) the method to be wrapped by NVTX range. If not provided (None),
+            the method will be inferred based on the object's class for Callable objects (like Transforms),
+            nn.Module objects (like Networks, Losses, etc.), and Dataloaders. Method resolve order is:
+            - forward()
+            - __call__()
+            - __next__()
+
+    """
+
+    name_counter: dict = defaultdict(int)
+
+    def __init__(self, name: Optional[str] = None, method: Optional[str] = None) -> None:
+        self.name = name
+        self.method = method
+
+    def __call__(self, obj: Any):
+        # Define the name to be associated to the range if not provided
+        if self.name is None:
+            name = type(obj).__name__
+            self.name_counter[name] += 1
+            self.name = f"{name}_{self.name_counter[name]}"
+
+        # Define the method to be wrapped if not provided
+        if self.method is None:
+            method_list = [
+                "forward",  # nn.Module
```

Comment:
Looks cool, I think nn.module has both forward and call, will it cause issues here?

Reply:
Thanks, @wyli! Good observation, that's why forward is before call in the list of methods. For `nn.Module`, it annotates `forward`, it's the method that we care about, right?

Reply:
I'm not sure, the documentation suggests call instead of using forward https://pytorch.org/docs/stable/generated/torch.nn.Module.html I think it depends on the granularity of measurement as well, maybe we need an option to report call/forward/both

Reply:
You're right, for `nn.Module` annotating call covers more computation than forward but the part that a user has control over  is forward. I checked the source code and the call overhead should be minimal so I'll change the code to annotate the call. 
Anyways, the option is already there to annotation any method and as many as one wants but this is about which method to annotate as default. I'll make them more specific for each type of components that we have in MONAI.

Reply:
Now it annotate both forward and call for `nn.Module` and append the method's name to the range's name to distinguish these two ranges.
https://github.com/Project-MONAI/MONAI/blob/69b9aae39dc416144c4663c3f7a31736f254079d/monai/utils/nvtx.py#L117